File: leetcode/path-sum-ii/path-sum-ii.java
// LeetCode #113 - Path Sum II
// https://leetcode.com/problems/path-sum-ii/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int targetSum) {
        List<List<Integer>> allPaths = new ArrayList<>();
        List<Integer> path = new ArrayList<>();

        var dfs = new ObjIntConsumer<TreeNode>() {
            @Override
            public void accept(TreeNode node, int target) {
                if (node == null) return;

                var subtarget = target - node.val;
                path.add(node.val);

                if (node.left != null || node.right != null) {
                    accept(node.left, subtarget);
                    accept(node.right, subtarget);
                } else if (subtarget == 0) {
                    allPaths.add(new ArrayList<>(path));
                }

                path.remove(path.size() - 1);
            }
        };

        dfs.accept(root, targetSum);
        return allPaths;
    }
}

File: leetcode/path-sum-ii/path-sum-ii.cpp
// LeetCode #113 - Path Sum II
// https://leetcode.com/problems/path-sum-ii/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    [[nodiscard]] static vector<vector<int>>
    pathSum(const TreeNode* root, int targetSum) noexcept;
};

vector<vector<int>>
Solution::pathSum(const TreeNode* const root, int targetSum) noexcept
{
    auto all_paths = vector<vector<int>>{};
    auto path = vector<int>{};

    const function<void(const TreeNode*)>
    dfs {[&](const TreeNode* const node) noexcept {
        if (!node) return;

        targetSum -= node->val;
        path.push_back(node->val);

        if (node->left || node->right) {
            dfs(node->left);
            dfs(node->right);
        } else if (targetSum == 0) {
            all_paths.push_back(path);
        }

        path.pop_back();
        targetSum += node->val;
    }};

    dfs(root);
    return all_paths;
}
